### PR TITLE
Add ViTDet

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

This PR adds part 1 of #25051, namely the ViTDet backbone, introduced in [Exploring Plain Vision Transformer Backbones for Object Detection](https://arxiv.org/abs/2203.16527).

Note that this PR only adds the backbone, hence there are no compatible checkpoints with the backbone-only. Those can only be added once either VitMatte or Mask R-CNN are added, both of which use VitDet as backbone.

### CODE REVIEW

File: src/transformers/models/vitdet/configuration_vitdet.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+            `"relu"`, `"selu"` and `"gelu_new"` are supported.
+        dropout_prob (`float`, *optional*, defaults to 0.0):
+            The dropout probability for all fully connected layers in the embeddings, encoder, and pooler.
+        initializer_range (`float`, *optional*, defaults to 0.02):
+            The standard deviation of the truncated_normal_initializer for initializing all weight matrices.
+        layer_norm_eps (`float`, *optional*, defaults to 1e-6):
+            The epsilon used by the layer normalization layers.
+        image_size (`int`, *optional*, defaults to 224):
+            The size (resolution) of each image.
+        pretrain_image_size (`int`, *optional*, defaults to 224):
```
Comment (amyeroberts) [inline src/transformers/models/vitdet/configuration_vitdet.py:60]: Just for my own understanding - what's the purpose of having `pretrain_image_size` and `image_size`? Looking at the modeling code - it seems this is so that the absolute embeddings are calculated relative to the images used for training vs the input images. Is that right?

File: src/transformers/models/vitdet/modeling_vitdet.py [hunk 2]
```diff
@@ -0,0 +1,892 @@
...
+
+        self.projection = nn.Conv2d(num_channels, hidden_size, kernel_size=patch_size, stride=patch_size)
+
+    def get_absolute_positions(self, abs_pos, has_cls_token, height, width):
+        """
+        Calculate absolute positional embeddings. If needed, resize embeddings and remove cls_token dimension for the
+        original embeddings.
+
+        Args:
+            abs_pos (`torch.Tensor`):
```
Comment (amyeroberts) [inline src/transformers/models/vitdet/modeling_vitdet.py:93]: `abs_pos` indicates the variable is the absolute position, rather than the embeddings. Let's go for something a bit clearer e.g.: 

```suggestion
            abs_pos_embeddings (`torch.Tensor`):
```

File: src/transformers/models/vitdet/modeling_vitdet.py [hunk 3]
```diff
@@ -0,0 +1,892 @@
...
+                Height of input image tokens.
+            width (`int`):
+                Width of input image tokens.
+
+        Returns:
+            Absolute positional embeddings after processing with shape (1, height, width, num_channels)
+        """
+        if has_cls_token:
+            abs_pos = abs_pos[:, 1:]
+        xy_num = abs_pos.shape[1]
```
Comment (amyeroberts) [inline src/transformers/models/vitdet/modeling_vitdet.py:107]: For clarifty and to match the shape info in the docstring

```suggestion
        num_position = abs_pos.shape[1]
```

File: src/transformers/models/vitdet/modeling_vitdet.py [hunk 4]
```diff
@@ -0,0 +1,892 @@
...
+            )
+
+        attention_probs = attention_scores.softmax(dim=-1)
+
+        hidden_state = (
+            (attention_probs @ values)
+            .view(batch_size, self.num_heads, height, width, -1)
+            .permute(0, 2, 3, 1, 4)
+            .reshape(batch_size, height, width, -1)
+        )
```
Comment (amyeroberts) [inline src/transformers/models/vitdet/modeling_vitdet.py:273]: Rather than do a one-liner that covers many lines, we can split this into separate steps. This will then be easier to debug if needed as we get a new result per-line (plus takes fewer lines!)

```suggestion
        hidden_state = attention_probs @ values
        hidden_state = hidden_state.view(batch_size, self.num_heads, height, width, -1)
        hidden_state = hidden_state.permute(0, 2, 3, 1, 4)
        hidden_state = hidden_state.reshape(batch_size, height, width, -1)
```

File: src/transformers/models/vitdet/modeling_vitdet.py [hunk 5]
```diff
@@ -0,0 +1,892 @@
...
+        outputs = (
+            (
+                hidden_state,
+                attention_probs.reshape(
+                    batch_size, self.num_heads, attention_probs.shape[-2], attention_probs.shape[-1]
+                ),
+            )
+            if output_attentions
+            else (hidden_state,)
+        )
```
Comment (amyeroberts) [inline src/transformers/models/vitdet/modeling_vitdet.py:285]: Same here - rather than try to do a one liner which forces using many lines we can split it up and use fewer lines overall + clearer 

```suggestion
        if output_attentions:
            attention_probs = attention_probs.reshape(
                batch_size, self.num_heads, attention_probs.shape[-2], attention_probs.shape[-1]
            )
            outputs = (hidden_state, attention_probs)
        else:
            outputs = (hidden_state,)
```

File: src/transformers/models/vitdet/modeling_vitdet.py [hunk 6]
```diff
@@ -0,0 +1,892 @@
...
+        return x
+
+
+class VitDetResBottleneckBlock(nn.Module):
+    """
+    The standard bottleneck residual block without the last activation layer. It contains 3 conv layers with kernels
+    1x1, 3x3, 1x1.
+    """
+
+    def __init__(
```
Comment (amyeroberts) [inline src/transformers/models/vitdet/modeling_vitdet.py:354]: Can be on one line

File: src/transformers/models/vitdet/modeling_vitdet.py [hunk 7]
```diff
@@ -0,0 +1,892 @@
...
+        self,
+        config,
+        in_channels,
+        out_channels,
+        bottleneck_channels,
+    ):
+        """
+        Args:
+            in_channels (int):
```
Comment (amyeroberts) [inline src/transformers/models/vitdet/modeling_vitdet.py:363]: * missing config
* code formatting needed for types

File: src/transformers/models/vitdet/modeling_vitdet.py [hunk 8]
```diff
@@ -0,0 +1,892 @@
...
+                Number of output channels.
+            bottleneck_channels (int):
+                Number of output channels for the 3x3 "bottleneck" conv layers.
+        """
+        super().__init__()
+        self.conv1 = nn.Conv2d(in_channels, bottleneck_channels, 1, bias=False)
+        self.norm1 = VitDetLayerNorm(bottleneck_channels)
+        self.act1 = ACT2FN[config.hidden_act]
+
+        self.conv2 = nn.Conv2d(
```
Comment (amyeroberts) [inline src/transformers/models/vitdet/modeling_vitdet.py:375]: Can be on one line

File: tests/models/vitdet/test_modeling_vitdet.py [hunk 9]
```diff
@@ -0,0 +1,306 @@
...
+    import torch
+    from torch import nn
+
+    from transformers import VitDetBackbone, VitDetModel
+
+
+if is_vision_available():
+    pass
+
+
```
Comment (amyeroberts) [inline tests/models/vitdet/test_modeling_vitdet.py:44]:
```suggestion
```

File: tests/models/vitdet/test_modeling_vitdet.py [hunk 10]
```diff
@@ -0,0 +1,306 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+""" Testing suite for the PyTorch ViTDet model. """
+
+
+import inspect
+import unittest
+
+from transformers import VitDetConfig
+from transformers.testing_utils import (
```
Comment (amyeroberts) [inline tests/models/vitdet/test_modeling_vitdet.py:22]: Can be on one line

File: tests/models/vitdet/test_modeling_vitdet.py [hunk 11]
```diff
@@ -0,0 +1,306 @@
...
+        config.output_hidden_states = True
+        config.output_attentions = self.has_attentions
+
+        # no need to test all models as different heads yield the same functionality
+        model_class = self.all_model_classes[0]
+        model = model_class(config)
+        model.to(torch_device)
+
+        print("Model class: ", model_class)
+
```
Comment (amyeroberts) [inline tests/models/vitdet/test_modeling_vitdet.py:267]:
```suggestion
```

File: src/transformers/models/vitdet/modeling_vitdet.py [hunk 12]
```diff
@@ -0,0 +1,892 @@
...
+
+
+class VitDetEncoder(nn.Module):
+    def __init__(self, config: VitDetConfig) -> None:
+        super().__init__()
+        self.config = config
+        depth = config.num_hidden_layers
+
+        # stochastic depth decay rule
+        dpr = [x.item() for x in torch.linspace(0, config.drop_path_rate, depth)]
```
Comment (amyeroberts) [inline src/transformers/models/vitdet/modeling_vitdet.py:556]: clearer var name

```suggestion
        drop_path_rate = [x.item() for x in torch.linspace(0, config.drop_path_rate, depth)]
```

File: src/transformers/models/vitdet/modeling_vitdet.py [hunk 13]
```diff
@@ -0,0 +1,892 @@
...
+    patch_height, patch_width = height + pad_height, width + pad_width
+
+    hidden_state = hidden_state.view(
+        batch_size, patch_height // window_size, window_size, patch_width // window_size, window_size, num_channels
+    )
+    windows = hidden_state.permute(0, 1, 3, 2, 4, 5).contiguous().view(-1, window_size, window_size, num_channels)
+    return windows, (patch_height, patch_width)
+
+
+def window_unpartition(windows, window_size, pad_hw, hw):
```
Comment (amyeroberts) [inline src/transformers/models/vitdet/modeling_vitdet.py:444]:
```suggestion
def window_unpartition(windows, window_size, pad_height_width, height_width):
```

**Review by amyeroberts — APPROVED** [review]
Thanks for adding this model! 

Very nice and clean PR :) Mostly just nits. The diff on the toctree will need to be resolved before merging.

File: src/transformers/models/vitdet/configuration_vitdet.py (hunk 1 above)
Comment (NielsRogge) [reply src/transformers/models/vitdet/configuration_vitdet.py:60]: Yes, for ViTMatte they set the `pretrain_image_size` to 224 and the `image_size` to 512 when leveraging ViTDet as backbone. 

The `pretrain_image_size` is used for initializing the absolute position embeddings, and the `image_size` is used for initializing the relative position embeddings. The authors use both absolute and relative position embeddings.

**Comment by NielsRogge** [comment]
@amyeroberts thanks for your review, I've addressed all the comments

File: docs/source/en/_toctree.yml
Comment (amyeroberts) [inline docs/source/en/_toctree.yml:1]: Reminder to resolve this. Once that's done we can merge in the PR

**Review by amyeroberts — COMMENTED** [review]
Thanks again for adding and iterating! 

There's still an issue with the toctree diff which needs to be addressed

File: docs/source/en/_toctree.yml
Comment (NielsRogge) [reply docs/source/en/_toctree.yml:1]: cc @ydshieh who knows the fix for this

Comment (ydshieh) [reply docs/source/en/_toctree.yml:1]: If you discard the change in this file of this PR,  rebase on main, adding back `ViTDet` entry to this file and run the style, it will be fine. 

A fix #25661 is merged into `main`.
